### PR TITLE
Add Transform component to the pipeline

### PR DESCRIPTION
**Pull request by deep-diver**
This PR includes the following **changes**:
- `tfx-pipeline.ipynb`: modified to demonstrate how `Transform` can be integrated into the pipeline
- `preprocessing.py`: handle low-resolution data preprocessing
- `preprocessing_full_res.py`: handle full-resolution data preprocessing
- `model.py` and `model_full_res.py`: replaced `parse_tfr` and `preprocess` steps from the previous version with transformed dataset from `Transform`
- `local_pipeline.py`: `Transform` component is added to the local pipeline
- `pipeline.py`: `Transform` component is added to the Vertex pipeline. Also, there is a condition to run `Transform` component in Dataflow
- `configs.py`: replace data source path to the re-generated TFRecords for the low-resolution dataset

**Note:** In order to use `Transform`, we need `StatisticsGen` and `SchemaGen`, so they are added to the pipeline as well.

**Note:** This code changes are verified to work on local and Vertex environment.

### CODE REVIEW

**Comment by sayakpaul** [comment]
@deep-diver the low-res PR is a dependency of this PR, right? Let's please hold off until it gets resolved.

**Comment by deep-diver** [comment]
@sayakpaul  

this is up for the review as well!

File: training_pipeline/models/model.py [hunk 1]
```diff
@@ -62,82 +71,32 @@ def serving_fn(string_input):
...
-    ).map(_parse_tfr, num_parallel_calls=tf.data.AUTOTUNE)
-
-    if is_train:
-        dataset = dataset.shuffle(batch_size * 2)
+    dataset = data_accessor.tf_dataset_factory(
+        file_pattern,
+        dataset_options.TensorFlowDatasetOptions(
+            batch_size=batch_size, label_key=_transformed_name(_LABEL_KEY)
+        ),
+        tf_transform_output.transformed_metadata.schema,
```
Comment (sayakpaul) [inline training_pipeline/models/model.py:93]: Let's add informative comments here:

* How `Transform` is being used.
* How it's inferring the TFRecord schema

Anything that would help the reader read the code better.

File: training_pipeline/models/model.py [hunk 2]
```diff
@@ -163,7 +122,9 @@ def _build_model(num_labels: int) -> tf.keras.Model:
         upsample(64, 3),  # 32x32 -> 64x64
     ]
 
-    inputs = tf.keras.layers.Input(shape=[128, 128, 3])
+    inputs = tf.keras.layers.Input(
+        shape=[128, 128, 3], name=_transformed_name(_IMAGE_KEY)
+    )
```
Comment (sayakpaul) [inline training_pipeline/models/model.py:127]: 128 should be defined as a meaningful variable.

File: training_pipeline/models/model.py [hunk 3]
```diff
@@ -187,7 +152,7 @@ def _build_model(num_labels: int) -> tf.keras.Model:
     model.compile(
         optimizer=Adam(learning_rate=_LR),
         loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
-        metrics=["accuracy"],
+        metrics=["sparse_categorical_accuracy"],
```
Comment (sayakpaul) [inline training_pipeline/models/model.py:155]: If you supply `accuracy` Keras will automatically infer the appropriate variant.

File: training_pipeline/models/model.py [hunk 4]
```diff
@@ -62,82 +71,32 @@ def serving_fn(string_input):
...
     _input_fn reads TFRecord files passed from the upstream 
-    TFX components such as ImportExampleGen. It calls _parse_tfr 
-    to parse each entry of the TFRecord files, then _preprocess
-    function is attached to resize the input image
+    TFX component, Transform. Assume the dataset is already 
+    transformed appropriately. 
 """
 
 
 def _input_fn(
```
Comment (sayakpaul) [inline training_pipeline/models/model.py:81]: Also denote what `_input_fn` is responsible of in this context.

File: training_pipeline/models/model.py (hunk 2 above)
Comment (sayakpaul) [reply training_pipeline/models/model.py:127]: Also, I was thinking maybe the model architecture should live in a separate module and we could then call this script something like `train.py`. Calling it `model.py` is vague IMO. 

WDYT?

**Review by sayakpaul — CHANGES_REQUESTED** [review]
@deep-diver looking good. I suggest we add a diagram that shows how the scripts in this PR are inter connected. That will help the readers navigate the code better. 

WDYT?

**Comment by deep-diver** [comment]
@sayakpaul 

addressed your comments!
thanks for the great comments. I really like them 👍🏼

**Comment by deep-diver** [comment]
@sayakpaul 

fixed the broken parts, and it is verified to be working properly.

<p align="center">
Screen Shot 2022-09-26 at 9 47 23 PM
</p>

**Review by sayakpaul — APPROVED** [review]
